### PR TITLE
Providers Axios Retry Mechanism

### PR DESCRIPTION
**Pull request by aruokhai**
### Objective
It is paramount for a service that makes multiple Http calls such as Esplora Provider to have an Error Retry Mechanism, which is what is being rectified by this pull request

### Changes
- Implemented a custom  error retry logic for http calls  with exponential delay logic 
-  Modified  how  error logs are thrown During Http calls

### Scope Of Change 
Crittical , as it affects the fetching of data.

Solves #17

### CODE REVIEW

File: package.json
Comment (theanmolsharma) [inline package.json:1]: As discussed, please remove this dependency and implement custom retry logic

File: src/configuration.model.ts [hunk 1]
```diff
@@ -57,6 +57,9 @@ class EsploraConfig {
         require_host: true,
     })
     url: string;
+    @IsInt()
```
Comment (theanmolsharma) [inline src/configuration.model.ts:60]: nit
```suggestion
    url: string;

    @IsInt()
```

**Review by theanmolsharma — CHANGES_REQUESTED** [review]

File: src/common/request.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        return false;
+    }
+
+    // Prevents retrying timed out & cancelled requests
+    if (CODE_EXCLUDE_LIST.includes(error.code)) {
+        return false;
+    }
+
+    return true;
+};
```
Comment (theanmolsharma) [inline src/common/request.ts:101]:
```suggestion
export const isNetworkError = (error): boolean => {
    return !(
        error.response ||
        !error.code ||
        ['ERR_CANCELED', 'ECONNABORTED'].includes(error.code)
    );
};
```

**Review by theanmolsharma — CHANGES_REQUESTED** [review]

File: src/common/request.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+};
+
+const exponentialDelay = (
+    retryNumber: number,
+    retryConfig: AxiosRetryConfig,
+): Promise<void> => {
+    const calculatedDelay = 2 ** retryNumber * retryConfig.delay;
+    const delay = calculatedDelay;
+    const randomSum = delay * 0.2 * Math.random(); // 0-20% of the delay
```
[the remark this answers is not in the document]
Comment (aruokhai) [reply src/common/request.ts:30]: It specifically is to add a little bit of jitter to make the calls less systematic which helps to avoids synchronised failures and distribute retries over time.  For example  If multiple request  attempt to retry a failed operation at the same exact time which is possible for esplora , they may all encounter the same underlying issue, leading to a cascading failure. Jitter introduces randomness into the retry timing, helping to spread out retries and reduce the likelihood of synchronized failures.

File: src/common/request.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+const axiosErrorResponse = (error: AxiosError) =>
+    error.response?.data || error.message;
+
+const exponentialDelay = (
+    retryNumber: number,
+    retryConfig: AxiosRetryConfig,
+): Promise<void> => {
+    const calculatedDelay = 2 ** retryNumber * retryConfig.delay;
+    const delay = calculatedDelay;
```
Comment (theanmolsharma) [inline src/common/request.ts:15]:
```suggestion
    const delay = 2 ** retryNumber * retryConfig.delay;
```

File: src/common/request.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+                await exponentialDelay(count, retryConfig);
+
+                continue;
+            }
+
+            logger.error(`unknown error encountered ${error}`);
+            throw error;
+        }
+    }
```
Comment (theanmolsharma) [inline src/common/request.ts:67]: Axios can sometimes throw [AggregateError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError), we should also handle them here.
doing something like
```ts
        if (error instanceof AggregateError) {
            logger.error(
                `Aggregate Error encountered: ${error.message}`,
                error.stack,
            );
            for (const cause of error.errors) {
                logger.error(`Cause: ${cause.message}`, cause.stack);
            }
        }
```

File: src/block-data-providers/esplora.provider.ts [hunk 6]
```diff
@@ -258,4 +224,8 @@ export class EsploraProvider
             url: `${this.baseUrl}/tx/${txid}`,
         });
     }
+
+    private async request(request_config: AxiosRequestConfig): Promise<any> {
+        return makeRequest(request_config, this.retryConfig, this.logger);
+    }
```
Comment (theanmolsharma) [inline src/block-data-providers/esplora.provider.ts:230]: Instead of this, we should directly call `makeRequest` wherever `request` is called

**Review by theanmolsharma — CHANGES_REQUESTED** [review]

File: src/common/request.ts (hunk 4 above)
Comment (aruokhai) [reply src/common/request.ts:15]: Funny how i didn't catch this :smile:

Comment (theanmolsharma) [reply src/common/request.ts]: nw dude, that's what reviews are for

**Pull request merged**